actors/migration/nv15/miner.go

```
@@ -3,79 +3,251 @@ package nv15
+			}
+
+			inSectors, err := miner6.LoadSectors(store, inRoot)
+			if err != nil {
+				return cid.Undef, xerrors.Errorf("failed to load inSectors: %w", err)
+			}
+
+			prevOutSectors, err := miner7.LoadSectors(store, prevOutRoot)
+			if err != nil {
+				return cid.Undef, xerrors.Errorf("failed to load prevOutSectors: %w", err)
+			}
+
+			for _, change := range diffs {
+				switch change.Type {
+				case amt.Remove:
+					if err := prevOutSectors.Delete(change.Key); err != nil {
+						return cid.Undef, xerrors.Errorf("failed to delete sector from prevOutSectors: %w", err)
+					}
+				case amt.Add:
+					fallthrough
+				case amt.Modify:
+					sectorNo := abi.SectorNumber(change.Key)
+					info, found, err := inSectors.Get(sectorNo)
+					if err != nil {
+						return cid.Undef, xerrors.Errorf("failed to get sector %d in inSectors: %w", sectorNo, err)
+					}
+
+					if !found {
+						return cid.Undef, xerrors.Errorf("didn't find sector %d in inSectors", sectorNo)
+					}
+
+					if err := prevOutSectors.Set(change.Key, migrateSectorInfo(*info)); err != nil {
+						return cid.Undef, xerrors.Errorf("failed to set migrated sector %d in prevOutSectors", sectorNo)
+					}
+				}
+			}
+
+			outArray = prevOutSectors.Array
+		} else {
+			// first time we're doing this, do all the work
+			outArray, err = adt.MakeEmptyArray(store, miner7.SectorsAmtBitwidth)
+			if err != nil {
+				return cid.Undef, xerrors.Errorf("failed to construct new sectors array: %w", err)
+			}
+
+			var sectorInfo miner6.SectorOnChainInfo
+			if err = inArray.ForEach(&sectorInfo, func(k int64) error {
+				return outArray.Set(uint64(k), migrateSectorInfo(sectorInfo))
+			}); err != nil {
+				return cid.Undef, err
+			}
+		}
+
+		outRoot, err := outArray.Root()
+		if err != nil {
+			return cid.Undef, xerrors.Errorf("error writing new sectors AMT: %w", err)
+		}
+
+		_ = cache.Write(MinerPrevSectorsInKey(minerAddr), inRoot)
```

Silently dropping the error here, because I wasn't sure whether concurrent access of the logger would be safe...anyone know?

---

Yes: https://github.com/filecoin-project/lotus/blob/dc26f3bd7453c515ca9aecc16fb9b3517784ef5e/chain/consensus/filcns/upgrades.go#L1205